### PR TITLE
Additional Materials, stereographic velocity plots, changed behaviour of rotate

### LINKED ISSUE #1: :star2: Feature request: Extend database of single-crystal and whole rock/aggregate elastic constants
---
# Feature request
### Summary:
  Extend database of single-crystal and whole rock/aggregate elastic constants, improve core materials code, and consolidate with Telewavesim

---

**Is your feature request related to a problem? Please describe.**
No - simply an extension.

**Describe the solution you'd like**
Extend the existing database to include (at least!) all those materials that are currently available in the `elast.py` module of the Telewavesim package. In tandem, build in some additional functionality to the `anisotropy.materials.Material` class itself. For example, calculating the Voigt/Reuss/Voigt-Reuss-Hill aggregate bulk/shear moduli (equations set out in Hill, 1952).

At this stage, it is worth considering what should be required for the elastic constants for a given material to be included. At a minimum, this should probably be two additional fields on the `anisotropy.materials.Material` class - one containing a material identifier, likely just a string, and the other containing a reference. This could either be the full reference string (probably best) or simply the DOI.

Comprehensive tests should also be written that explore the functions in `anisotropy.materials.core.py` and some simple tutorial examples could be written that describe how to load a material from the database, how to mix an aggregate material, how to water the single-crystal elastic constants down with their isotropic component, how to define a custom material that is not in the database (and maybe some instructions on how to contribute it).

To-do list
------------
- [x] Expand elastic database to include all those minerals that are available in Telewavesim - 4bcf0f3
  - [x] Implement generalised methods to calculate the bulk/shear moduli (in the cases of both isotropic and generally anisotropic elastic tensors) - 32f4354
  - [x] Add `material_id` and `material_reference` attributes to `Material` class - 57d1fdf
- [x] Write a function to calculate the isotropic component of a general elastic stiffness tensor (Browaeys and Chevrot, 2004) - 98fe159
- [ ] Write tests
  - [x] Elastic constant averaging functions
  - [x] Bulk/shear moduli functions
  - [ ] Calculate polarisation - this one I definitely need to spend a little time understanding. Sketch vector diagram once solved.
  - [x] Extensive test for the conversion between Voigt/full 4th rank tensor form - test all 81 elements of the tensor!
- [x] Create tutorials - 839e959
  - [x] Loading a material
  - [x] Mixing an aggregate
  - [x] Watering down single-crystal tensors with their isotropic component
  - [x] Defining custom materials (not in the database)

Something to do in a separate PR would be to test this module in-situ in the Telewavesim package, to ensure that all requirements are met there and this module can then become the single source for elastic tensor representation.

### LINKED ISSUE #4: Material.rotate() should rotate self.C or return Material, instead of returning C
Currently, `Material.rotate(alpha, beta, gamma, **kwargs) ` returns a numpy array of the rotated tensor. In my opinion, it would be more intuitive if it would either:

a) return nothing, but write the rotated tensor straight to `self.C`, or
b) return a copy of Material with the rotated tensor

imho (a) is more intuitive, but there may be reasons to prefer (b).

I'm happy to implement either.

### PR DESCRIPTION
**Pull request by wasjabloch**
This pull request includes a new features and some changes:

- Material.rotate now returns save and may act in place
- new material hexagonal1_C, which implements anisotropy definition of [PyRaysum](https://github.com/paudetseis/PyRaysum)
- New method Material.plot_velocity()
- Fixed bug regarding complex values in velocities calculated from Christoffel matrix
- averaging of materials does not require volume fractions to add up to unity.

Sorry for some missed `rebase`s.

### CODE REVIEW

File: anisotropy/materials/core.py [hunk 1]
```diff
@@ -302,6 +302,160 @@ def rotate(self, alpha, beta, gamma, order=[1, 2, 3], mode="extrinsic"):
...
+                Show minimum / maximum values in plot
+            cmap: (str)
+                Color map to use to fraw colors
+            ax: (matpltlib.Axes)
+                Draw axis in external figure
+            fig_kw: (dict)
+                kwargs passed to created figure (ignored when ax is used)
+        """
+
+        # import matplotlib.pyplot as mp
```
Comment (hemmelig) [inline anisotropy/materials/core.py]:
```suggestion
```

File: anisotropy/utils/core.py [hunk 2]
```diff
@@ -154,3 +154,11 @@ def wrapper(*args, **kwargs):
...
         return wrapper
     return inner_function
+
+def sample_sphere(n):
+    """n even samples over a half sphere"""
+    # https://stackoverflow.com/questions/9600801/evenly-distributing-n-points-on-a-sphere
+    idx = np.arange(0, n, dtype=float) + 0.5
+    incs = ((np.pi - np.arccos(idx / n)) * 180 / np.pi) % 90
+    azis = ((np.pi * (1 + 5**0.5) * idx) * 180 / np.pi) % 360
+    return incs, azis
```
Comment (hemmelig) [inline anisotropy/utils/core.py]:
```suggestion
    return incs, azis

```

File: anisotropy/materials/database.py [hunk 3]
```diff
@@ -648,6 +677,150 @@ def _olivine():
...
+#C36          0.2200     
+#C44         71.6600     
+#C45          0.0600   
+#C46          0.3800         
+#C55         64.4900
+#C56          0.1300   
+#                       
+#C66         68.2700     
+
+
```
Comment (hemmelig) [inline anisotropy/materials/database.py:823]:
```suggestion


```

**Review by hemmelig — COMMENTED** [review]
Everything seems ok to me - I'll install this branch and run the tests on my local machine before approving the PR. Great to have some plotting functionality!

**Comment by hemmelig** [comment]
Tests currently failing:

- No `mplstereonet` installed - solved by running `pip install mplstereonet`. Should at least add this as a formal dependency, or an optional dependency and add information to install instructions + catch the potentially failing import. My only concern with `mplstereonet` is it appears to (as of 2020) no longer be actively supported. There is a warning emitted by `matplotlib`:

```
Testing the function 'plot_velocity'...
        Test 1 - testing basic plot...
/opt/miniconda3/envs/anisotropy/lib/python3.9/site-packages/mplstereonet/stereonet_axes.py:17: PendingDeprecationWarning: Overriding `Axes.cla` in StereonetAxes is pending deprecation in 3.6 and will be fully deprecated in favor of `Axes.clear` in the future. Please report this to the 'mplstereonet.stereonet_axes' author.
```

which is slightly concerning - a very simple fix (find + replace all instances of `.cla` with `.clear`), but if it is never merged then it will potentially break things a year or so. I will submit the fix to the main repo, regardless.

After installing `mplstereonet`:

```
	Test 4 - testing a suite of cases where one
		 expects an exception to be raised...
F
======================================================================
FAIL: test_voigt_reuss_hill_averaging (__main__.TestMaterials)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/conor/Documents/b_research/b_active_projects/b_anisotropy/AnisotroPy/tests/test_materials.py", line 113, in test_voigt_reuss_hill_averaging
    self.assertRaises(ValueError, materials.voigt_reuss_hill_average,
AssertionError: ValueError not raised by voigt_reuss_hill_average
```

This was due to the `ValueError` no longer being raised when the volume fractions did not sum to 1. Instead, the volume fractions are automatically re-normalised to 1. This seems fine in principle to me, but I think should be accompanied by a warning message, since the values may have been mis-entered. I've added the following:

```

    # Validate volume fraction sums to unity
    sumfrac = np.sum(volume_fractions)
    if sumfrac != 1:
        volume_fractions = np.divide(volume_fractions, sumfrac)
        warnings.warn(
            "Volume fractions did not sum to unity - re-normalising."
            "\nDouble-check your chosen volume fractions are correct."
        )

```

as well as the appropriate import.

**Comment by hemmelig** [comment]
I've made some changes locally - could you give me permission to push those changes to the PR branch on your fork? I believe this SO answer should explain how - https://stackoverflow.com/a/49854013

Cheers! Once done, I think this PR should be good to merge!

**Comment by hemmelig** [comment]
I made a minor git blunder before (force pushed an old branch to the PR, overwriting a number of commits), but I believe I've managed to undo it. I'll double check everything is as we expect before merging.

**Comment by wasjabloch** [comment]
I unfortunately cannot see and review your most recent changes to come around the tests I caused to fail (sorry, I created the pull-request last thing before the holidays and was somewhat rushed).  I only see three tiny commits which I can't approve ("cannot approve your own pull-request") and a force-push that appears to _undo_ my commits.

I made you a collaborator on my fork, so you should be able to push directly into it (the checkbox suggested in the SO answer has already been checked).

Can you take it from here?

**Comment by wasjabloch** [comment]
> My only concern with `mplstereonet` is it appears to (as of 2020) no longer be actively supported.

One could probably extract the `mplstereonet` code that implements the wulff- / schmidt-net projections and leave `mplstereonet` for good. However, to me `mplstereonet` appears actively maintained (just merged a PR 4 days ago) and seems as a tractable dependency for the development version of `anisotropy`.

**Review by hemmelig — APPROVED** [review]
Seems I've managed to undo the git mistake and can now approve this PR - can you just confirm you see your previous commits in the above? Tests pass and all of your changes are present for me!

**Comment by hemmelig** [comment]
Yes - it does appear there is still some activity on mplstereonet (though the recent PR merged was the first in 2 years). Fine for now, can resolve if/when there is a need.

**Comment by hemmelig** [comment]
@wsja Hey - I'd like to merge these changes, but want to just confirm that it looks ok on your end. Thanks!

**Comment by wasjabloch** [comment]
Yes, looks good to me. Thanks for catching my glitches.

**Comment by hemmelig** [comment]
Great!

**Pull request merged**